executor/analyze.go

```diff
@@ -64,54 +59,78 @@ func (e *AnalyzeExec) Close() error {
 
 // Next implements the Executor Next interface.
 func (e *AnalyzeExec) Next() (*Row, error) {
-	for _, src := range e.Srcs {
-		ae := src.(*AnalyzeExec)
-		var count int64 = -1
-		var sampleRows []*ast.Row
-		var colNDVs []int64
-		if len(ae.colIDs) != 0 {
-			rs := &recordSet{executor: ae.Srcs[len(ae.Srcs)-1]}
+	for i, src := range e.Srcs {
+		switch src.(type) {
+		case *XSelectTableExec:
 			var err error
-			count, sampleRows, colNDVs, err = CollectSamplesAndEstimateNDVs(rs, len(ae.colIDs))
+			if i < e.pkCount {
+				err = e.analyzePK(src.(*XSelectTableExec))
+			} else {
+				err = e.analyzeColumns(src.(*XSelectTableExec))
+			}
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
-		}
-		columnSamples := rowsToColumnSamples(sampleRows)
-		var pkRS ast.RecordSet
-		if ae.pkID != 0 {
-			offset := len(ae.Srcs) - 1
-			if len(ae.colIDs) != 0 {
-				offset--
+		case *XSelectIndexExec:
+			err := e.analyzeIndex(src.(*XSelectIndexExec))
+			if err != nil {
+				return nil, errors.Trace(err)
 			}
-			pkRS = &recordSet{executor: ae.Srcs[offset]}
-		}
-		idxRS := make([]ast.RecordSet, 0, len(ae.idxIDs))
-		for i := range ae.idxIDs {
-			idxRS = append(idxRS, &recordSet{executor: ae.Srcs[i]})
-		}
-		err := ae.buildStatisticsAndSaveToKV(count, columnSamples, colNDVs, idxRS, pkRS)
-		if err != nil {
-			return nil, errors.Trace(err)
 		}
 	}
 	return nil, nil
 }
 
-func (e *AnalyzeExec) buildStatisticsAndSaveToKV(count int64, columnSamples [][]types.Datum, colNDVs []int64, idxRS []ast.RecordSet, pkRS ast.RecordSet) error {
+func (e *AnalyzeExec) analyzePK(exec *XSelectTableExec) error {
+	statBuilder := &statistics.Builder{
+		Ctx:        exec.ctx,
+		TblInfo:    exec.tableInfo,
+		Count:      -1,
+		NumBuckets: defaultBucketCount,
+		PkRecords:  &recordSet{executor: exec},
+		PkID:       exec.Columns[0].ID,
+	}
+	err := e.buildStatisticsAndSaveToKV(statBuilder)
```

Comment:
we can export some functions in statistic package instead of call a same function.
Like `BuildColumns(colsInfo, tableInfo)`, `BuildIndex(idxInfo,tableInfo)`,  build*** can dump to TiKV directly. We call them concurrently. After all goroutines have done, sleep 2 lease